### ISSUE TITLE
Make indent and quote style configurable.

### ISSUE THREAD

**Issue by bolinfest** [bazelbuild/buildifier#34]
To get this to follow Buck's conventions, I had to make some small changes:

https://github.com/bolinfest/buildifier/commit/68b6d08b1e2dd2887ae1d786e20a33abf3bf98c9

It would be nice if these things were configurable, as we would prefer not to maintain a fork.

**Comment by pmbethe09** [comment]
I saw your indent flag PR: https://github.com/bazelbuild/buildifier/pull/35
That seems fine since the default configuration can be unchanged.

If you want to do the same for those other features, that sounds fine to me, as long as the defaults are preserved as is.  Too many BUILD files inside Google would be broken otherwise.

**Comment by bolinfest** [comment]
Yep, now that I got legal signoff to contribute, I'll put up a PR for this!

**Comment by pmbethe09** [comment]
And as with the other PR, give existing defaults in the core package for library-based usage.  Thanks.

**Comment by laurentlb** [comment]
Hi,

My preference is to avoid this kind of settings, and have everyone use the same style. Tools based on Buildifier (e.g. Buildozer) are going to modify files using the standard formatting and ignore your options. Would it be an option for you to adopt Buildifier default style?

**Comment by bolinfest** [comment]
This is the style we have been using at Facebook. As the person who
introduced the style, I'd rather keep what we have and avoid updating
hundreds of our build files just to flip the quotes and lengthen the lines.

**Comment by bolinfest** [comment]
Obviously we could also fork, but I would prefer not to do that because
then it's unlikely that we'll contribute anything else back going forward.

**Comment by bolinfest** [comment]
Incidentally, Buildozer does not appear to be a public tool. Is it going to be? If so, when? And on what timeline?

**Comment by bolinfest** [comment]
Without knowing the details of what Buildozer does, if using it were really important to us, presumably we could use `buildifier` to reformat our files in the Google format, run Buildozer to do whatever it does, and then run `buildifier` again with our settings to convert them back.

**Comment by pmbethe09** [comment]
buildozer was open sourced in this repo a month ago
https://github.com/bazelbuild/buildifier/commit/91f1a2a9ca5ace0b06d86153d7521492348e54bd
with a bug-fix Dec 15
https://github.com/bazelbuild/buildifier/commit/736c577fa275bbc4176f19d0c654cb3f93b12062

**Comment by bolinfest** [comment]
@pmbethe09 Ah, thanks for the pointer. https://www.google.com/search?q=site%3Abazel.build%20buildozer turned up nothing and https://github.com/bazelbuild/buildozer doesn't exist, which were the two places I looked.

**Comment by laurentlb** [comment]
Hi,

If you need these changes, we can approve them. I first wanted to know more about your use-case.

But a few points:
 - Buildifier is strongly inspired by gofmt. By design, gofmt has no configuration option, and I'm sure any PR will be rejected. You can read about it here: https://talks.go-zh.org/2015/gofmt-en.slide#2

 - The goal is to get consistency. We've released Buildifier early, because we feel that it's important to get everyone use the same style. Compare the situation in C++ (every project has its own style) with Go (everyone uses gofmt). When we wrote Buildifier, we reformatted every single BUILD file at Google to match the new formatting. Although it was a large change, it was well worth it.

 - As soon as you start using a new formatter, you have to reformat all the existing code - or live with the inconsistencies. Even with your PR, many of your files will have to be updated. So if the files are going to be changed, maybe it's worth adopting the same style as the outside world?

- I'm afraid some people will start using the options and set their own preferences. That would defeat one of our goals.

If you tell me the options are needed to match the Buck style, maybe we could have a single option --style=bazel or --style=buck? What do you think? That would make it easier for other Buck users to follow your style.

Thanks!

**Comment by bolinfest** [comment]
@laurentlb Thanks for providing such detailed context! Let me talk to the team and see whether they would be open to just changing the style to match what buildifier does today.

There's no reason that we couldn't use the current indent and quoting settings, but one option that is required for Buck today when using buildifier is `-buildifier_disable=label`. Specifically, Buck does not allow `//foo/bar` to be used as a shorthand for `//foo/bar:bar`. This was an intentional design decision that I made because I found this inconsistency confusing.

When Google uses buildifier internally, does it always prefer the shorthand version? Are you willing to commit to supporting `-buildifier_disable=label` long-term? Given your references to `gofmt` and its deliberate absence of configuration, I'm surprised this option exists under the hood today.

**Comment by ulfjack** [comment]
Yes, it always prefers the shorthand version. The current format is not set in stone, and we'd be happy to discuss changes or give more details about why it was chosen to be this way.

I would prefer if we could agree on a single style, and if we have to reformat all of our BUILD files for that, then so be it. After all, we have made such changes in the past. That said, if we can't agree on a single style and decide to support multiple styles, we'll need to find a way to make sure that there's still a mechanism to enforce a consistent style, at least per project.

**Comment by laurentlb** [comment]
I don't think anyone has used buildifier_disable - except at the beginning, when we were updating our full code base to the new formatting.

Yes, we can definitely support 'buildifier_disable=label' in the long-term. Is there any other buildifier transformation that can break BUCK files?

**Comment by bolinfest** [comment]
@ulfjack @laurentlb OK, if we can commit to supporting the `-buildifier_disable=label` flag, then I think we're almost there. The other issue is being able to customize this file:

https://github.com/bazelbuild/buildifier/blob/master/tables/tables.go

I don't know if you had a general plan for making this extensible, since presumably people who define their own rules via Skylark should be able to add entries to these tables, as well.

We would probably want to add an option that allows reading these values from a JSON config file or something. I suspect you would prefer to keep the built-in ones defined in code so you don't have to pay the cost of parsing JSON at runtime, but we could live with that.

Let me know if/what your preference is for configuration here and I'll put up a PR for it.

**Comment by bolinfest** [comment]
@ulfjack 

> Yes, it always prefers the shorthand version. The current format is not set in stone, and we'd be happy to discuss changes or give more details about why it was chosen to be this way.

The justification for Buck's format is here: https://buckbuild.com/concept/build_target.html. In particular, take note of the **Command-line Protip** section. Specifically, we are strict on the format within build files, but loose in the format on the command line (which I think is appropriate because that's where people are actually typing build targets/labels by hand).

For example, you have to type this in Bazel:

    bazel build //buildifier

but you cannot type this:

    bazel build buildifier/

but if this were Buck, you could do this:

    buck build buildifier/

The last one is particularly handy because it is shell-tab-complete friendly.

**Comment by ulfjack** [comment]
We have a shell completion script. :-) I'd also like to point out that you can type

    bazel build buildifier

But we don't currently ignore trailing slashes in the CLI. We could change that, I suppose.

**Comment by bolinfest** [comment]
OK, there's one open question, which I think got buried in the above, so I'll restate:

https://github.com/bazelbuild/buildifier/issues/34#issuecomment-273507961

Basically, in what way would you be willing to make `tables.go` configurable? I'm volunteering to write the code.

**Comment by pmbethe09** [comment]
If we follow Laurent's suggestion of keeping just -style=bazel (default)
and -style=buck, then I would think a 'buckTables.go' would make sense.

**Comment by bolinfest** [comment]
If you're on board with `-style=buck`, then I can do that. (I thought we were trying to avoid options, if that's ultimately possible.)

I don't know how often `tables.go` changes. If it's a lot, it would be nice to not have to rebuild to update it. I expect that to release `buildifier` internally, I'm going to have to build it for Linux, OS X, and Windows, and then put a Python wrapper around it to select the right binary, so doing a release is going to be a little tedious, so I'd like to do so sparingly.

Is there a plan for `buildifier` and end-user defined rules in Skylark?

**Comment by pmbethe09** [comment]
We have been discussing a 'skylarkifier', but only discussion at this time.

**Comment by bolinfest** [comment]
@pmbethe09 Maybe my mental model is wrong, but I mean that if I write a custom rule for Bazel in Skylark and then I use that rule in a `BUILD` file, but I want to control how my usages are formatted by `buildifier`, what's the story there? You would need to change `buildifier` rather than a `skylarkifier`, right?

If it were possible to pass some additional config to `buildifier`, then it would help both the Buck case and the external-person-extending-Bazel-via-Skylark case whereas `-style=buck` just helps Buck. I'm open to either, I just wanted to throw it out there.

**Comment by pmbethe09** [comment]
On Fri, Jan 20, 2017 at 1:17 PM, bolinfest <notifications@github.com> wrote:

> @pmbethe09 <https://github.com/pmbethe09> Maybe my mental model is wrong,
> but I mean that if I write a custom rule for Bazel in Skylark and then I
> use that rule in a BUILD file, but I want to control how my usages are
> formatted by buildifier, what's the story there? You would need to change
> buildifier rather than a skylarkifier, right?
>

Ah, yes of course.  Sorry if I misread your earlier comment.  What would
you do differently at the call-site of a skylark rule?


> If it were possible to pass some additional config to buildifier, then it
> would help both the Buck case and the external-person-extending-Bazel-via-Skylark
> case whereas -style=buck just helps Buck. I'm open to either, I just
> wanted to throw it out there.
>

I think the idea is to try and make people choose between one of these 2
styles, rather than allow more configuration options -- and thus just
compiling in the buck alternative styling and providing the flag.  Unless I
misunderstood what Ulf and Laurent were suggesting.

**Comment by bolinfest** [comment]
@pmbethe09 For a custom Skylark rule, I might need to configure which arguments can or cannot be alpha-sorted. For example, it's important that the `srcs` for `genrule` is **not** alpha-sorted, right? Ideally, that setting would be communicated to buildifier somehow.

Nevertheless, I'm on board with putting up a PR to implement `-style=buck`, which would include our own `tables.go`. I don't believe we introduce new built-in build rules all that often, so hopefully keeping `buck-tables.go` up-to-date won't require too much churn.

Though I'm curious if the public version of buildifier has a sanitized version of `tables.go` that excludes internal-only Google rules (presumably [Copybara](https://github.com/google/copybara) or [MOE](https://github.com/google/MOE) is used to perform that sanitization). If we ended up in a situation where we had internal-only Buck rules that we wanted to keep private for some reason, then we would still have to fork buildifier internally to support formatting such rules. Whereas if buildifier accepted a command-line argument to specify a JSON file or something that defined this configuration, then we would be sure that we would not need to fork.

**Comment by pmbethe09** [comment]
Ahh...  that makes sense.   As you suspected, internally at Google, we swap
in an augmented version of tables.go via Copybara.

A pluggable json config sounds fine to me.

**Comment by ulfjack** [comment]
How will buildifier find the config?

**Comment by pmbethe09** [comment]
The default config will be baked in, to use with an alternate tables
configuration, pass  -tables <path-to-file>

**Comment by bolinfest** [comment]
OK, I'm going to work on the `-tables` option. I believe that once we have that (and if we switch to double-quoted strings and four space indents), then instead of `-style=buck`, we can just implement everything we need for Buck via:

```
#!/bin/bash
BUILDIFIER_DIFF="diff -u" buildifier -tables=buck_tables.json --buildifier_disable=label "$@"
```

Any tips for cross-platform binary releases of tools built in Go? I assumed I would have to compile buildifier for each of Mac, Linux, and Windows, and then bundle it with a Python script or something that selects the appropriate binary for the platform. Is there a canonical way to do this?

**Comment by bolinfest** [comment]
Huzzah, here is a PR based on what we've discussed on this thread: https://github.com/bazelbuild/buildifier/pull/45.

**Comment by mwoehlke-kitware** [comment]
> Compare the situation in C++ (every project has its own style) with Go (everyone uses gofmt)

Code style is an area where everyone has a very strong, and different, opinion. Python (PEP-8) tries to follow the Go model, but you still have people that think some (all?) of PEP-8 is asinine and absolutely *will not* follow it. Given a sufficiently large user base, I'm fairly confident that non-uniformity is inevitable.

The reason C++ has so many coding styles is because it has an *enormous* user base. K&R style is a thing, too. You might want to keep in mind how well *that* caught on.

That said... I'm actually mildly in favor of *trying* to create a One True Style; I just think that will only "stretch" so far...

It also helps to have very, very few holes in the formatting tools. Unfortunately (#109) this is not true of buildifier.

> We'll need to find a way to make sure that there's still a mechanism to enforce a consistent style, at least per project

The usual way is having a `.buildifier` file in the project root, much like `bazel.rc`.

**Comment by pmbethe09** [comment]
Actually we will push for One True Style and except as necessary to also work for BUCK not allow configurability.  When we release support for .bzl I am sure there will be those against some of the choices, but internally we have a formatter for almost everything and people have learned to code, format, submit and not worry if they would have made slightly different formatting choices.

**Comment by laurentlb** [comment]
Thanks for the feedback!

> Given a sufficiently large user base

It's not only that. There also (mostly?) about habits and legacy code. By releasing Buildifier early and by using the same consistent style throughout the documentation, it's very likely that users will follow the same style. PEP-8 was written 10 years after Python was created, and tooling came even later. I think that makes a huge difference.

Apart for Buck, I don't expect users to have a strong reason to follow a different style.
